sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/_evaluators/_document_retrieval/_document_retrieval.py

```diff
@@ -0,0 +1,323 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+
+import json
+import math
+import operator
+from itertools import starmap
+from azure.ai.evaluation._evaluators._common import EvaluatorBase
+from typing import Dict, List, TypedDict, Optional
+
+
+DocumentLabel = TypedDict(
+    'DocumentLabel',
+    {
+        "document_id": str,
+        "label": float
+    }
+)
+
+
+DocumentRetrievalMetrics = TypedDict(
+    'DocumentRetrievalMetrics',
+    {
+        "ndcg@3": float,
+        "ndcg@3_result": bool,
+        "xdcg@3": float,
+        "xdcg@3_result": bool,
+        "fidelity": float,
+        "fidelity_result": bool,
+        "top1_relevance": float,
+        "top1_relevance_result": bool,
+        "top3_max_relevance": float,
+        "top3_max_relevance_result": bool,
+        "holes": int,
+        "holes_result": bool,
+        "holes_ratio": float,
+        "holes_ratio_result": bool,
+        "total_retrieved_documents": int,
+        "total_groundtruth_documents": int
+    }
+)
+
+
+class DocumentRetrievalEvaluator:
+    """
+    Calculate document retrieval metrics, such as NDCG, XDCG, Fidelity and Top K Relevance.
+    """
+    def __init__(self, groundtruth_min: int = 0, groundtruth_max: int = 4, groundtruth_step: int = 1, threshold: Optional[dict] = None):
+        super().__init__()
+        self.k = 3
+        self.xdcg_discount_factor = 0.6
+        self.groundtruth_min = groundtruth_min
+        self.groundtruth_max = groundtruth_max
+        self.groundtruth_step = groundtruth_step
+        
+        # The default threshold for metrics where higher numbers are better.
+        default_threshold = {
+            "ndcg@3": 0.5,
+            "xdcg@3": 0.5,
+            "fidelity": 0.5,
+            "top1_relevance": 50,
+            "top3_max_relevance": 50,
+            "total_retrieved_documents": 50,
+            "total_groundtruth_documents": 50
+        }
+
+        # Ideally, the number of holes should be zero.
+        self._threshold_holes = {
+            "holes": 0,
+            "holes_ratio": 0
+        }
+
+        if threshold is None:
+            threshold = {}
+
+        elif not isinstance(threshold, dict):
+            raise TypeError(
```

Comment:
Please use the Evaluation Exception class for it

Reply:
Updated!